pkg/gameservers/controller.go

```diff
@@ -334,8 +334,24 @@ func (ext *Extensions) creationMutationHandlerPod(review admissionv1.AdmissionRe
 
 	ext.baseLogger.WithField("pod.Name", pod.Name).Debug("creationMutationHandlerPod")
 
-	// TODO: We need to deal with case of multiple and mixed type ports before enabling the feature gate.
-	pod.Spec.Containers[1].Ports[0].ContainerPort = pod.Spec.Containers[1].Ports[0].HostPort
+	annotation, ok := pod.ObjectMeta.Annotations[agonesv1.PassthroughPortAssignmentAnnotation]
+	if !ok {
+		ext.baseLogger.WithField("pod.Name", pod.Name).Debug("creationMutationHandlerPod, " + agonesv1.PassthroughPortAssignmentAnnotation + ": " + pod.ObjectMeta.Annotations[agonesv1.PassthroughPortAssignmentAnnotation])
+	}
+
+	passthroughPortAssignmentMap := make(map[string][]int)
+	if err := json.Unmarshal([]byte(annotation), &passthroughPortAssignmentMap); err != nil {
+		return review, errors.Wrapf(err, "could not unmarshal annotation %q (value %q)", passthroughPortAssignmentMap, annotation)
+	}
+
+	for _, p := range pod.Spec.Containers {
+		if containerPassthrough, ok := passthroughPortAssignmentMap[p.Name]; ok {
+			for i := range containerPassthrough {
+				i := containerPassthrough[i]
+				p.Ports[i].ContainerPort = p.Ports[i].HostPort
```

Comment:
So, this is subtle but I think your logic here is off. In particular, in gameserver.go you're recording the index within the Ports array, but here you're indexing into container.Ports by the _absolute_ index, vs the index within that container. You can see this in your new unit text where the annotation records `containerFour: [5]`, but that's index 5 into the GameServer Ports array, not index 5 into the container Ports array.

To fix this, I'd recommend two things:
* In gamserver.go, instead of adding an index variable to the range over the GameServer ports, record the index before the line `c.Ports = append(c.Ports, cp)`, I.e. by `portIdx := len(c.Ports)`.
* Test the multi-container case here as well - in fact if you mirror the container/port structure in the test above you just added, it'll make for a good symmetric test.

Reply:
Ohhhh interesting, good catch! I made the changes and now i'm using the container port index instead and updated unit test to reflect this mirror in the multi container test, thanks!!